### DISCUSSION TITLE
Keyboard input support from Bitlocker login

### DISCUSSION THREAD

**Discussion by tomjohnstone** [hrvach/deskhop#122] [Ideas]
It would be nice to be able to use my keyboard from the Bitlocker login screen, just after the bios screen. I can plug a keyboard in and it will work on this screen, but I can't switch over from my other machine. ~~I wonder if this is due to caps lock being re-mapped to F24 in the new RC and not being able to switch via the keyboard, since there is no mouse support on the Bitlocker login screen~~ _edit: I confirmed that caps lock does not work either_. That said when I try to move the mouse from my other machine (A), the light for machine (B) on the Deskhop does illuminate.

Setup:
Machine A is primary machine with mouse & keyboard connected via USB hub to keyboard USB port on Deskhop
Machine B has BitLocker

**Comment by tomjohnstone** [comment]
I also tried disabling fast boot mode, but that didn't work either:

https://answers.microsoft.com/en-us/windows/forum/all/keyboard-not-working-in-bitlocker/fda262a6-ed41-403e-a423-7a19305b22ef

**Comment by hrvach** [comment]
I think I know what might work - a dedicated endpoint for the keyboard without using report ID. Maybe try that? Or build it without the mouse support and don't send report ID just for test and tell me if it works...

I'm sure there is some way we could make a workaround for such small / embedded environments.

**Reply by tomjohnstone** [reply]
Sure thing, I'm unfamiliar with HID so if you can point me in the right direction I can take a look at coding up either of those options to experiment with. Thanks!

**Reply by tomjohnstone** [reply]
This is my attempt so far at option 1. It is not registering any input from neither the keyboard or mouse however:

https://github.com/tomjohnstone/deskhop/pull/2/files

**Comment by hrvach** [comment]
I think adding a tud_hid_set_protocol_cb callback and detecting if protocol is HID_PROTOCOL_BOOT, and then stripping the ID in this case for keyboard reports and disabling other reports would be a quick'n'dirty workaround.

**Reply by tomjohnstone** [reply]
tud_hid_set_protocol_cb only receives an instance and protocol. Where do I get at the ID to strip it, from this callback?

**Reply by hrvach** [reply]
@tomjohnstone you don't need to know the ID to strip it, I think just setting it to 0 would suffice for tinyusb. I'm working on this, we definitely need proper boot protocol support as a client and I hope to have something ready soon. Just have to find something to test on...

**Reply by tomjohnstone** [reply]
Would your bios be a good testing ground? When you get something working I'm happy to test on my Bitlocker login too. Cheers!

**Comment by Joonetic** [comment]
Hi there.  Sorry for the bumping an old post but I am also not able to use the keyboard when trying to type in my pw for bitlocker. Happy to be a tester.

**Reply by hrvach** [reply]
I just tried setting up bitlocker and I can't re-create this. It works just fine for me (but only numbers are allowed to be typed, since the pin is numeric). Just to confirm, neither letters nor numbers work?

**Reply by Joonetic** [reply]
Hi yep both letters and numbers do not work.

**Reply by hrvach** [reply]
Can you please share more details - which windows version? I'll try to install something in a VM and recreate... if I'm able to recreate, fixing shouldn't be an issue but with my Win11 I wasn't able to somehow - it just worked.

**Reply by Joonetic** [reply]
Hi the laptop I was trying was win 10. I tried another (same model, Dell latitude 7420) on win 11 and still didn't work. But what's interesting is I tried a precision 7550 and that worked. I copied the exact bios settings and it's still not working on the 7420 which is odd. The only difference is the bios on the 7550 is the "old school" look and the other models have the newer more modern gui. If I connect the keyboard directly bitlocker works though.

Edit: I also tried a few more models (7560, 7670,and 5770) and none worked. I went through again and made sure the bios settings were the same so I'm not sure what I'm missing.

**Comment by mglushko** [comment]
I had a go at fixing this, was having the same issue. Seems to be working for me. #356
